csrc/ir/nodes.cpp

```diff
@@ -389,9 +389,81 @@ UnaryOp::UnaryOp(IrBuilderPasskey passkey, UnaryOpType type, Val* out, Val* in)
 
 std::vector<PolymorphicValue> UnaryOp::evaluate(
     const ExpressionEvaluator& ee,
-    const std::vector<PolymorphicValue>& inputs) const {
+    std::unordered_map<const Val*, PolymorphicValue>& known_values) const {
   using namespace PolymorphicValue_functions;
-  const auto& in = inputs.at(0);
+
+  // If the UnaryOp is CastOp, check if the preceding pattern of
+  // operators matches with matmul (MmaOp(Broadcast (A), Broadcast(B)) -> Cast)
+  // or matmul + bias (BinaryOp::Add (MmaOp(Broadcast (A), Broadcast(B),
+  // Broadcast(bias)) -> Cast) If not, evaluate UnaryOp::CastOp along with the
+  // other types by evaluating the immediate input.
+
+  auto has_matmul_and_bias = [](Val* in) -> bool {
+    if (auto* binary = dynamic_cast<BinaryOp*>(in->definition())) {
+      if (binary->getBinaryOpType() == BinaryOpType::Add) {
+        if (binary->input(0)->definition()->isA<MmaOp>()) {
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+
+  if ((getUnaryOpType() == UnaryOpType::Cast) &&
+      input(0)->definition() != nullptr) {
```

Comment:
We should probably use NVF_ERROR to check that definition is not null instead of skipping this branch in that unexpected case.

Reply:
That is a valid case. For eg: a standalone castOp. In that case, it will be evaluated regularly with the other UnaryOp. Why do you think this case should not occur?

Reply:
Oh you're right. Sorry I was just confused. Thanks for the explanation.